File: perl5/sync_clawhub_git.pl
#!/usr/bin/perl
# sync_clawhub_git.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:scripts/sync_clawhub_git.py
# Erzeugt: 2026-08-20 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use File::Path qw(make_path remove_tree);
use File::Copy qw(copy);
use File::Find;
use Digest::SHA qw(sha256_hex);
use Getopt::Long;
use Time::Piece;

# Konfiguration
my $CLAWHUB_DIR = "/home/openclaw/.openclaw/workspace/skills";
my $GIT_DIR = "/home/openclaw/.openclaw/workspace/git/skills";
my $BACKUP_DIR = "/home/openclaw/.openclaw/workspace/backups/sync";
my $LOG_FILE = "/home/openclaw/.openclaw/workspace/logs/sync-agent.log";

# Erstelle Verzeichnisse
make_path($GIT_DIR, { verbose => 0 });
make_path($BACKUP_DIR, { verbose => 0 });
make_path((split '/', $LOG_FILE)[0..$#_-1], { verbose => 0 });

# Logging
sub log_message {
    my ($message, $level) = @_;
    $level //= "INFO";
    my $timestamp = localtime->strftime('%Y-%m-%d %H:%M:%S');
    my $entry = "[$timestamp] [$level] $message\n";
    print $entry;
    open(my $fh, '>>', $LOG_FILE) or die "Could not open file '$LOG_FILE' $!";
    print $fh $entry;
    close $fh;
}

# Validierung
sub validate_skill {
    my ($skill_dir) = @_;
    my $skill_md = "$skill_dir/SKILL.md";
    unless (-e $skill_md) {
        log_message("Validation failed: " . (split '/', $skill_dir)[-1] . " missing SKILL.md", "ERROR");
        return 0;
    }
    return 1;
}

# Backup
sub create_backup {
    my ($source, $skill_name) = @_;
    my $timestamp = localtime->strftime('%Y%m%d_%H%M%S');
    my $backup_path = "$BACKUP_DIR/${skill_name}_${timestamp}";

    # Backup verzeichnis löschen falls es existiert
    if (-e $backup_path) {
        eval {
            remove_tree($backup_path);
            log_message("Removed existing backup: $backup_path");
        };
        if ($@) {
            log_message("Failed to remove existing backup $backup_path: $@", "ERROR");
            return 0;
        }
    }

    eval {
        copy_recursive($source, $backup_path);
        log_message("Backup created: $backup_path");
        return 1;
    };
    if ($@) {
        log_message("Backup failed: $@", "ERROR");
        return 0;
    }
}

# Rekursives Kopieren
sub copy_recursive {
    my ($src, $dst) = @_;
    opendir(my $dh, $src) or die "Cannot open directory '$src': $!";
    mkdir $dst or die "Cannot create directory '$dst': $!" unless -e $dst;
    while (readdir $dh) {
        next if ($_ eq '.' || $_ eq '..');
        next if ($_ eq '.git'); # Ignoriere .git Verzeichnis
        my $src_path = "$src/$_";
        my $dst_path = "$dst/$_";
        if (-d $src_path) {
            copy_recursive($src_path, $dst_path);
        } else {
            copy($src_path, $dst_path) or die "Copy failed: $!";
        }
    }
    closedir $dh;
}

# Hash-Vergleich
sub get_file_hash {
    my ($file_path) = @_;
    open(my $fh, '<', $file_path) or die "Cannot open file '$file_path': $!";
    binmode($fh);
    my $hash = sha256_hex(do { local $/; <$fh> });
    close $fh;
    return $hash;
}

# Sync Richtung ClawHub → Git
sub sync_to_git {
    my ($skill_name, $dry_run) = @_;
    $dry_run //= 1;
    my $source = "$CLAWHUB_DIR/$skill_name";
    my $target = "$GIT_DIR/$skill_name";

    return 0 unless validate_skill($source);

    # Backup vor Änderungen (nur wenn target existiert)
    if (!$dry_run && -e $target) {
        create_backup($target, $skill_name) or return 0;
    }

    # Änderungen erkennen
    my @changes;
    find(sub {
        return if $_ eq '.git';
        my $rel_path = $File::Find::name;
        $rel_path =~ s/^$source//;
        $rel_path =~ s/^\///;
        return unless -f $_;
        my $tgt_file = "$target/$rel_path";
        if (!-e $tgt_file) {
            push @changes, "ADD $rel_path";
        } elsif (get_file_hash($_) ne get_file_hash($tgt_file)) {
            push @changes, "UPDATE $rel_path";
        }
    }, $source);

    # Dry-Run Report
    if ($dry_run) {
        log_message("DRY-RUN: $skill_name - " . scalar(@changes) . " changes");
        for my $change (@changes) {
            log_message("  $change");
        }
        return 1;
    }

    # Echte Synchronisation
    log_message("SYNC: $skill_name - Applying " . scalar(@changes) . " changes");
    if (-e $target) {
        remove_tree($target);
    }
    copy_recursive($source, $target);
    log_message("SYNC: $skill_name - Complete");
    return 1;
}

# Sync Richtung Git → ClawHub
sub sync_to_clawhub {
    my ($skill_name, $dry_run) = @_;
    $dry_run //= 1;
    my $source = "$GIT_DIR/$skill_name";
    my $target = "$CLAWHUB_DIR/$skill_name";

    return 0 unless validate_skill($source);

    # Backup vor Änderungen (nur wenn target existiert)
    if (!$dry_run && -e $target) {
        create_backup($target, $skill_name) or return 0;
    }

    # Änderungen erkennen (gleiche Logik wie oben)
    my @changes;
    find(sub {
        return if $_ eq '.git';
        my $rel_path = $File::Find::name;
        $rel_path =~ s/^$source//;
        $rel_path =~ s/^\///;
        return unless -f $_;
        my $tgt_file = "$target/$rel_path";
        if (!-e $tgt_file) {
            push @changes, "ADD $rel_path";
        } elsif (get_file_hash($_) ne get_file_hash($tgt_file)) {
            push @changes, "UPDATE $rel_path";
        }
    }, $source);

    # Dry-Run Report
    if ($dry_run) {
        log_message("DRY-RUN: $skill_name - " . scalar(@changes) . " changes");
        for my $change (@changes) {
            log_message("  $change");
        }
        return 1;
    }

    # Echte Synchronisation
    log_message("SYNC: $skill_name - Applying " . scalar(@changes) . " changes");
    if (-e $target) {
        remove_tree($target);
    }
    copy_recursive($source, $target);
    log_message("SYNC: $skill_name - Complete");
    return 1;
}

# Hauptfunktion
sub main {
    my ($skill, $direction, $dry_run, $force);
    GetOptions(
        "skill=s"     => \$skill,
        "direction=s" => \$direction,
        "dry-run"     => \$dry_run,
        "force"       => \$force,
    ) or die "Error in command line arguments\n";

    die "Missing --skill argument\n" unless defined $skill;
    die "Missing --direction argument\n" unless defined $direction;
    die "Invalid direction. Use 'to-git' or 'to-clawhub'\n"
        unless $direction eq 'to-git' || $direction eq 'to-clawhub';

    log_message("Starting sync: $skill ($direction)");

    my $success;
    if ($direction eq 'to-git') {
        $success = sync_to_git($skill, $dry_run);
    } else {
        $success = sync_to_clawhub($skill, $dry_run);
    }

    unless ($success) {
        log_message("Sync failed", "ERROR");
        exit 1;
    }

    log_message("Sync completed");
}

main() if !caller;

File: javascript/sync_clawhub_git.js
#!/usr/bin/env node
// sync_clawhub_git.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:scripts/sync_clawhub_git.py
// Erzeugt: 2026-08-20 durch ABSTRACTIONS_MANAGER.py

/**
 * Bidirektionale ClawHub ↔ Git Synchronisation
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { program } = require('commander');

// Konfiguration
const CLAWHUB_DIR = path.join('/home/openclaw/.openclaw/workspace/skills');
const GIT_DIR = path.join('/home/openclaw/.openclaw/workspace/git/skills');
const BACKUP_DIR = path.join('/home/openclaw/.openclaw/workspace/backups/sync');
const LOG_FILE = path.join('/home/openclaw/.openclaw/workspace/logs/sync-agent.log');

// Erstelle Verzeichnisse
[CLAWHUB_DIR, GIT_DIR, BACKUP_DIR, path.dirname(LOG_FILE)].forEach(dir => {
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
});

// Logging
function log(message, level = "INFO") {
  const timestamp = new Date().toISOString().replace('T', ' ').slice(0, 19);
  const entry = `[${timestamp}] [${level}] ${message}`;
  console.log(entry);
  fs.appendFileSync(LOG_FILE, entry + '\n');
}

// Validierung
function validateSkill(skillDir) {
  /** Prüft Skill-Struktur - SKILL.md required, scripts/ optional */
  const skillMdPath = path.join(skillDir, "SKILL.md");
  if (!fs.existsSync(skillMdPath)) {
    log(`Validation failed: ${path.basename(skillDir)} missing SKILL.md`, "ERROR");
    return false;
  }
  return true;
}

// Backup
function createBackup(source, skillName) {
  /** Erstellt Backup eines Skills */
  const timestamp = new Date().toISOString().replace(/[-:]/g, '').replace('T', '_').slice(0, 15);
  const backupPath = path.join(BACKUP_DIR, `${skillName}_${timestamp}`);
  
  // Backup verzeichnis löschen falls es existiert
  if (fs.existsSync(backupPath)) {
    try {
      fs.rmSync(backupPath, { recursive: true, force: true });
      log(`Removed existing backup: ${backupPath}`);
    } catch (e) {
      log(`Failed to remove existing backup ${backupPath}: ${e.message}`, "ERROR");
      return false;
    }
  }
  
  try {
    copyRecursiveSync(source, backupPath);
    log(`Backup created: ${backupPath}`);
    return true;
  } catch (e) {
    log(`Backup failed: ${e.message}`, "ERROR");
    return false;
  }
}

// Hilfsfunktion für rekursives Kopieren
function copyRecursiveSync(src, dest) {
  const stats = fs.statSync(src);
  if (stats.isDirectory()) {
    if (!fs.existsSync(dest)) {
      fs.mkdirSync(dest, { recursive: true });
    }
    fs.readdirSync(src).forEach(child => {
      if (child !== '.git') {
        copyRecursiveSync(path.join(src, child), path.join(dest, child));
      }
    });
  } else {
    fs.copyFileSync(src, dest);
  }
}

// Hash-Vergleich
function getFileHash(filePath) {
  /** SHA256-Hash einer Datei */
  const hash = crypto.createHash('sha256');
  const stream = fs.createReadStream(filePath);
  
  return new Promise((resolve, reject) => {
    stream.on('data', chunk => hash.update(chunk));
    stream.on('end', () => resolve(hash.digest('hex')));
    stream.on('error', reject);
  });
}

// Sync Richtung ClawHub → Git
async function syncToGit(skillName, dryRun = true) {
  /** Synchronisiert ClawHub Skill zu Git */
  const source = path.join(CLAWHUB_DIR, skillName);
  const target = path.join(GIT_DIR, skillName);
  
  if (!validateSkill(source)) {
    return false;
  }
  
  // Backup vor Änderungen (nur wenn target existiert)
  if (!dryRun && fs.existsSync(target)) {
    createBackup(target, skillName);
  }
  
  // Änderungen erkennen
  const changes = [];
  
  function walkDir(currentPath, relativeBase) {
    const items = fs.readdirSync(currentPath);
    items.forEach(item => {
      if (item === '.git') return;
      
      const fullPath = path.join(currentPath, item);
      const stat = fs.statSync(fullPath);
      const relPath = path.relative(relativeBase, currentPath);
      const targetItemPath = path.join(target, relPath, item);
      
      if (stat.isDirectory()) {
        walkDir(fullPath, relativeBase);
      } else {
        if (!fs.existsSync(targetItemPath)) {
          changes.push(`ADD ${path.join(relPath, item)}`);
        } else {
          // Für DRY-RUN können wir hier vereinfacht vergleichen
          // In Produktion sollte getFileHash verwendet werden
        }
      }
    });
  }
  
  if (fs.existsSync(source)) {
    walkDir(source, source);
  }
  
  // Für echte Implementierung mit Hash-Vergleich:
  async function collectChangesWithHash(srcBase, tgtBase) {
    const changesList = [];
    
    async function walkWithHash(currentSrc, relBase) {
      const items = fs.readdirSync(currentSrc);
      for (const item of items) {
        if (item === '.git') continue;
        
        const srcPath = path.join(currentSrc, item);
        const stat = fs.statSync(srcPath);
        const relPath = path.relative(relBase, currentSrc);
        const tgtPath = path.join(tgtBase, relPath, item);
        
        if (stat.isDirectory()) {
          await walkWithHash(srcPath, relBase);
        } else {
          if (!fs.existsSync(tgtPath)) {
            changesList.push(`ADD ${path.join(relPath, item)}`);
          } else {
            const srcHash = await getFileHash(srcPath);
            const tgtHash = await getFileHash(tgtPath);
            if (srcHash !== tgtHash) {
              changesList.push(`UPDATE ${path.join(relPath, item)}`);
            }
          }
        }
      }
    }
    
    if (fs.existsSync(srcBase)) {
      await walkWithHash(srcBase, srcBase);
    }
    return changesList;
  }
  
  const actualChanges = await collectChangesWithHash(source, target);
  
  // Dry-Run Report
  if (dryRun) {
    log(`DRY-RUN: ${skillName} - ${actualChanges.length} changes`);
    actualChanges.forEach(change => log(`  ${change}`));
    return true;
  }
  
  // Echte Synchronisation
  log(`SYNC: ${skillName} - Applying ${actualChanges.length} changes`);
  if (fs.existsSync(target)) {
    // Entferne Zielverzeichnis-Inhalt außer .git
    const targetItems = fs.readdirSync(target);
    targetItems.forEach(item => {
      if (item !== '.git') {
        fs.rmSync(path.join(target, item), { recursive: true, force: true });
      }
    });
    copyRecursiveSync(source, target);
  } else {
    copyRecursiveSync(source, target);
  }
  log(`SYNC: ${skillName} - Complete`);
  return true;
}

// Sync Richtung Git → ClawHub
async function syncToClawhub(skillName, dryRun = true) {
  /** Synchronisiert Git Skill zu ClawHub */
  const source = path.join(GIT_DIR, skillName);
  const target = path.join(CLAWHUB_DIR, skillName);
  
  if (!validateSkill(source)) {
    return false;
  }
  
  // Backup vor Änderungen (nur wenn target existiert)
  if (!dryRun && fs.existsSync(target)) {
    createBackup(target, skillName);
  }

  // Änderungen erkennen (gleiche Logik wie oben)
  async function collectChangesWithHash(srcBase, tgtBase) {
    const changesList = [];
    
    async function walkWithHash(currentSrc, relBase) {
      const items = fs.readdirSync(currentSrc);
      for (const item of items) {
        if (item === '.git') continue;
        
        const srcPath = path.join(currentSrc, item);
        const stat = fs.statSync(srcPath);
        const relPath = path.relative(relBase, currentSrc);
        const tgtPath = path.join(tgtBase, relPath, item);
        
        if (stat.isDirectory()) {
          await walkWithHash(srcPath, relBase);
        } else {
          if (!fs.existsSync(tgtPath)) {
            changesList.push(`ADD ${path.join(relPath, item)}`);
          } else {
            const srcHash = await getFileHash(srcPath);
            const tgtHash = await getFileHash(tgtPath);
            if (srcHash !== tgtHash) {
              changesList.push(`UPDATE ${path.join(relPath, item)}`);
            }
          }
        }
      }
    }
    
    if (fs.existsSync(srcBase)) {
      await walkWithHash(srcBase, srcBase);
    }
    return changesList;
  }
  
  const actualChanges = await collectChangesWithHash(source, target);
  
  // Dry-Run Report
  if (dryRun) {
    log(`DRY-RUN: ${skillName} - ${actualChanges.length} changes`);
    actualChanges.forEach(change => log(`  ${change}`));
    return true;
  }

  // Echte Synchronisation
  log(`SYNC: ${skillName} - Applying ${actualChanges.length} changes`);
  if (fs.existsSync(target)) {
    // Entferne Zielverzeichnis-Inhalt
    const targetItems = fs.readdirSync(target);
    targetItems.forEach(item => {
      fs.rmSync(path.join(target, item), { recursive: true, force: true });
    });
    copyRecursiveSync(source, target);
  } else {
    copyRecursiveSync(source, target);
  }
  log(`SYNC: ${skillName} - Complete`);
  return true;
}

// Hauptfunktion
async function main() {
  program
    .description('Bidirektionaler Sync ClawHub ↔ Git')
    .requiredOption('--skill <name>', 'Skill name')
    .option('--direction <dir>', 'Sync direction', 'to-git') // Standardwert hinzugefügt
    .option('--dry-run', 'Nur Änderungen anzeigen')
    .option('--force', 'Ohne Backup')
    .action(async (options) => {
      const { skill, direction, dryRun, force } = options;
      
      if (!['to-git', 'to-clawhub'].includes(direction)) {
        console.error('Error: --direction must be either "to-git" or "to-clawhub"');
        process.exit(1);
      }
      
      log(`Starting sync: ${skill} (${direction})`);
      
      let success;
      if (direction === 'to-git') {
        success = await syncToGit(skill, dryRun);
      } else {
        success = await syncToClawhub(skill, dryRun);
      }
      
      if (!success) {
        log("Sync failed", "ERROR");
        process.exit(1);
      }
      
      log("Sync completed");
    });
    
  program.parse();
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});
